### PR TITLE
MNG-6049: Adds version range result filter behaviour

### PR DESCRIPTION
**Pull request by barthel**
The discussion on issue [MNG-3092](https://issues.apache.org/jira/browse/MNG-3092) shows the seriously needs of different kinds of version range resolving in Maven.

I know about the risk on changing the resolving mechanism in Maven.
I always had the backward compatibility in mind, if I have something changed.
I added some basic tests for the default version range resolver before I've started with the refactoring.

The idea behind is the ```DependencySelector``` suggested by @jvanzyl in [comment](https://issues.apache.org/jira/browse/MNG-3092?focusedCommentId=14417976&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-14417976):
> So allow a pluggable DependencySelector to Aether is the answer here. But we need to find a way to expose it.

But without compromise the dependency resolving algorithm itself.

This pull request contains the basic tests and a **less invasive** solution based on a filter that works on resolved version range results.

The actual version range resolving algorithm **wasn't changed** and the internal used default filter implementation don't filter anything.

The goal:
Provide a hook for Maven extensions/plugins to change the list of resolved version range results as required.

These changes are only placed in _maven-aether-provider_.

### CODE REVIEW

**Comment by barthel** [comment]
This PR replaced the PR (#68) and based on master and re-created branch MNG-3092. This branch contains the cherry picked b6d7269.

**Comment by barthel** [comment]
@jvanzyl @michael-o : Ready for easier review.

**Comment by michael-o** [comment]
@ChristianSchulte Maybe `Rage` is intentional ;-)

**Comment by barthel** [comment]
@michael-o There is a typo and not a hidden statement. :-D

**Comment by barthel** [comment]
@michael-o Your comment on 'old' commit b6d72699c8b0089351f797e217d635424c1ae5a2:
> Don't! Always through a NPE if null is passed.

:+1: But, I'm oriented on the current used pattern like in #setRepositoryEventDispatcher or #setSyncContextFactory.

> Right, see my ongoing effort on this: https://issues.apache.org/jira/browse/MNG-5649

Should I adapt the other methods? It's not the original intention of this PR.
Maybe I create a branch and/or new commit with these changes only.

**Comment by barthel** [comment]
@ChristianSchulte Typo fixed. Thanks for looking into it.

**Comment by barthel** [comment]
This solution could also be usable for [MNG-5353](https://issues.apache.org/jira/browse/MNG-5353).

File: .gitignore [hunk 1]
```diff
@@ -13,3 +13,5 @@ out/
 /bootstrap
 /dependencies.xml
 .java-version
+nb-configuration.xml
+
```
Comment (michael-o) [inline .gitignore]: Not related to the PR, please remove.

File: maven-aether-provider/src/main/java/org/apache/maven/repository/internal/DefaultVersionRangeResolver.java [hunk 2]
```diff
@@ -152,6 +154,16 @@ public DefaultVersionRangeResolver setRepositoryEventDispatcher( RepositoryEvent
         return this;
     }
 
+    public DefaultVersionRangeResolver setVersionRangeResultFilter( VersionRangeResultFilter versionRangeResultFilter )
+    {
+        if ( versionRangeResultFilter == null )
+        {
+            throw new IllegalArgumentException( "version range result filter has not been specified" );
+        }
```
Comment (michael-o) [inline maven-aether-provider/src/main/java/org/apache/maven/repository/internal/DefaultVersionRangeResolver.java]: Please use Commons Lang as I have done it recently on other modules in Maven.

**Comment by barthel** [comment]
@michael-o Your review comments are integrated.
.gitignore cleaned up.
Apache Commons Lang used instead of ``IllegalArgumentException``.
Please remove version property for commons-lang when merging into the master.

**Comment by michael-o** [comment]
@barthel Thanks for the update. If you rebase against current master you don't need that property at all. I have added Commons Lang to dependency management already.

**Comment by michael-o** [comment]
@jvanzyl Can a I merge this PR into master? Are you OK with that?

**Comment by barthel** [comment]
@michael-o @jvanzyl Let's go merging? :-)

**Comment by barthel** [comment]
Can I do anything to support the merge?

**Comment by michael-o** [comment]
@jvanzyl Are you saying that you dislike to change the default behavior, i.e., no merge?  @barthel Did you run the ITs before and after your change?

**Comment by barthel** [comment]
The default behavior is ``do nothing``. The ITs doesn't show any errors at this modifications.

**Comment by jvanzyl** [comment]
Then I don’t see an issue with merging it.

> On Oct 27, 2015, at 4:38 AM, barthel <notifications@github.com> wrote:
> 
> The default behavior is do nothing. The ITs doesn't show any errors at this modifications.
> 
> —
> Reply to this email directly or view it on GitHub.
> 

Thanks,

Jason

----------------------------------------------------------
Jason van Zyl
Founder, Takari and Apache Maven
http://twitter.com/jvanzyl
http://twitter.com/takari_io
---------------------------------------------------------

A man enjoys his work when he understands the whole and when he
is responsible for the quality of the whole

 -- Christopher Alexander, A Pattern Language

**Comment by barthel** [comment]
I would like to see this changes in the next Maven (3.3.10) release.
@michael-o @jvanzyl Please don't forget to merge :-)

**Comment by michael-o** [comment]
@barthel GitHub tells me that this branch has conflicts. Can you kindly check if rebase on top of master if necessary?

**Comment by barthel** [comment]
@michael-o see my comment Oct 19:
> Please remove version property for commons-lang when merging into the master.

That's the rework you suggest with working on [https://issues.apache.org/jira/browse/MNG-5649].

**Comment by michael-o** [comment]
@barthel See my inline comments. Then I will take it from there.

**Comment by barthel** [comment]
POM file adapted. Use parent version 3.3.10-SNAPSHOT and remove ```commons-lang``` dependency.

**Comment by michael-o** [comment]
Thanks, I will test and merge tomorrow.

**Comment by michael-o** [comment]
@barthel Git gives me:

>  git merge --no-commit barthel/MNG-3092
Auto-merging maven-aether-provider/src/main/java/org/apache/maven/repository/internal/DefaultVersionRangeResolver.java
CONFLICT (content): Merge conflict in maven-aether-provider/src/main/java/org/apache/maven/repository/internal/DefaultVersionRangeResolver.java
Automatic merge failed; fix conflicts and then commit the result.

Can you resolve and push again to this branch?

**Comment by barthel** [comment]
Organize import in ```DefaultVersionRangeResolver``` and squash commit.

**Comment by michael-o** [comment]
There are still some issues with the branch. You did not replay on top of master. It is creating a merge commit which I'd like to avoid. Some other issues will be commented inline.

**Comment by jvanzyl** [comment]
Just looking through here and you also switch these to use JSR330 annotations instead of the plexus annotations. Any additions I'd like to see use JSR330 which is what we typically do now. I have a large branch which is a mass conversion of everything to JSR330 and it will help if when rebasing that I don't have to convert components again while merging which is what I've been doing for the last year with this branch. I don't mind doing this myself, just let me know.

**Comment by barthel** [comment]
@michael-o Indent size and inside issues solved. Commits squashed and amend on last commit in branch.
What can I do to avoid the merge commit? If I merge the master into this branch a lot more commits are involved here.

**Comment by barthel** [comment]
@jvanzyl Did you mean ```@Requirement``` and ```@Component```?
I'm following the coding style found in the classes around.

Please point me to the replacements for these annotations.

**Comment by michael-o** [comment]
@barthel While I am not a Git expert. I extract a patch/stash fo the work. Branch master, apply stash and commit again. [This](http://stackoverflow.com/q/16358418/696632) is probably worth reading.

**Comment by jvanzyl** [comment]
@barthel I mean using the JSR330 equivalents for @Component and @Requirement. For @Component we replace them with @Named and @Singleton used together at the class level, and we replace @Requirement with @Inject at the field or constructor level. Generally I think using private final fields and using @Inject on the constructor is the better option. One we do not have with Plexus annotations. An example of a newer component that uses these are:

https://github.com/apache/maven/blob/master/maven-core/src/main/java/org/apache/maven/extension/internal/CoreExportsProvider.java

**Comment by jvanzyl** [comment]
@barthel I ask this to be done because with some recent changes in Sisu I actually have a chance to land my JSR330 branch which entirely removes Plexus from Maven. @mcculls added support for @PostConstruct and @PreDestroy which allows me to replace all uses of the Disposable interface which gets me another chunk of the way.

**Comment by jvanzyl** [comment]
@barthel if you want me to take your PR and make the JSR330 changes I'm happy to give it a whirl. Just let me know.

**Comment by barthel** [comment]
I'm busy the next few days. But I will give it a try if these changes have a chance to merge in 3.3.10.

**Comment by jvanzyl** [comment]
We're looking at the new year for the next core release so you have time. I might have my JSR330 changes done by then. I have 15 super nasty uses of Plexus to get around but doable over the holidays.

**Comment by michael-o** [comment]
Can we make any progress here?

**Comment by pSub** [comment]
I would appreciate a merge, too.

**Comment by jvanzyl** [comment]
If there are integration tests don't wait for me.

**Comment by michael-o** [comment]
@jvanzyl There are unit tests in the commit. Isn't that enough? Otherwise, I'd ask @barthel for it.

**Comment by jvanzyl** [comment]
For a feature like that there should be an integration test as well.

**Comment by michael-o** [comment]
This is what I have expected you to say ;-)
@barthel or anyone else, can you craft up an IT for that?

**Comment by barthel** [comment]
@jvanzyl , @michael-o : Sorry for the delay. I'm busy right now. I will try to do the change to Maven 3.4 (and JSR330) and add a integration test. But not before friday this week.

**Comment by barthel** [comment]
It's an old branch. So I merged the actual master into it and use Maven 3.4.0-SNAPSHOT instead.

Only the last commit reflects the MNG-3092 solution.

@jvanzyl : I'll start to work on JSR 330.

**Comment by jvanzyl** [comment]
Not critical, I can add the annotations later. More important are integration tests if you haven't started those yet. If you have, great!

**Comment by barthel** [comment]
@jvanzyl , @michael-o  : I start working on the integration tests. It's more complicated than I thought.

The default handling works but I have difficulties to inject the filter into the ```DefaultVersionRangeResolver```.

**Comment by jvanzyl** [comment]
I suggest looking at more recent integration tests which are similar to yours where there is a default implementation and new implementations can be used. Take a look at MavenITmng5591WorkspaceReader.

**Comment by barthel** [comment]
@jvanzyl : Thanks for your tip. Now it works. :-D

**Comment by barthel** [comment]
Use JSR 330 annotation instead of the Plexus annotation.

**Comment by barthel** [comment]
Integration test pull request: https://github.com/apache/maven-integration-testing/pull/14

**Comment by barthel** [comment]
@jvanzyl @michael-o : Read to merge.

**Comment by michael-o** [comment]
This commit is huge and has a lot of merge noise. Do you think you can squeeze it down to your changes only?

**Comment by barthel** [comment]
@michael-o Sorry I've no idea how to remove the merge 'noise' from this really old branch. The only change of mine is ```28fd43d```.

**Comment by dinoboy197** [comment]
@barthel - I think all you should need to do is to rebase your feature branch (MNG-3092) against upstream/master.

Something like this should do it (assuming that in your git clone, upstream is configured to point to apache/maven and origin is configured to point to barthel/maven):
* `git fetch upstream`
* `git checkout MNG-3092`
* `git rebase upstream/master`
* `git push -f origin MNG-3092`

**Comment by barthel** [comment]
@dinoboy197 : Fantastic. Thanks for your help. After reading your post, it sounds so logical.

@michael-o , @jvanzyl : Now the merge comes true? :-D

**Comment by michael-o** [comment]
@barthel much but I won't be able to take a closer at it before next week. @jvanzyl if you can do it sooner, just go again and do it.

**Comment by barthel** [comment]
@ChristianSchulte : Sounds interesting. I'll take a look on it at the end of this week.

**Comment by ChristianSchulte** [comment]
@barthel: See [c31be833a6f8d9266990a708efe624d09fa06dec](https://github.com/ChristianSchulte/maven/commit/c31be833a6f8d9266990a708efe624d09fa06dec)

**Comment by dinoboy197** [comment]
@michael-o @jvanzyl @barthel - just checking in to see if any progress has been made on reviewing this.

**Comment by barthel** [comment]
@dinoboy197: Sorry, I haven't had time to look at @ChristianSchulte suggestion until now.
@jvanzyl @michael-o: Suggestion: Merge this solution. I'll evaluate the ```ContextualSnapshotVersionFilter``` way in a separate task. If this solution is a equivalent or a better one, I'll start a new pull request for use the way @ChristianSchulte  suggested and removing my solution.

WDYT?

**Comment by barthel** [comment]
I would like to look at ```ContextualSnapshotVersionFilter``` solution, if the Aether code was moved back to Apache Maven.

File: maven-aether-provider/src/main/java/org/apache/maven/repository/internal/DefaultVersionRangeResolver.java [hunk 3]
```diff
@@ -63,14 +64,16 @@
...
 /**
  * @author Benjamin Bentmann
  */
 @Named
-@Component( role = VersionRangeResolver.class )
+@Singleton
+@Component( role = VersionRangeResolver.class, hint = "default" )
 public class DefaultVersionRangeResolver
-    implements VersionRangeResolver, Service
+        implements VersionRangeResolver, Service
```
Comment (michael-o) [inline maven-aether-provider/src/main/java/org/apache/maven/repository/internal/DefaultVersionRangeResolver.java]: Unrelated change, revert!

File: maven-aether-provider/src/main/java/org/apache/maven/repository/internal/DefaultVersionRangeResolver.java [hunk 4]
```diff
@@ -88,19 +91,27 @@
...
+
     public DefaultVersionRangeResolver()
     {
         // enable default constructor
     }
 
     @Inject
     DefaultVersionRangeResolver( MetadataResolver metadataResolver, SyncContextFactory syncContextFactory,
-                                 RepositoryEventDispatcher repositoryEventDispatcher, LoggerFactory loggerFactory )
+            RepositoryEventDispatcher repositoryEventDispatcher, LoggerFactory loggerFactory,
```
Comment (michael-o) [inline maven-aether-provider/src/main/java/org/apache/maven/repository/internal/DefaultVersionRangeResolver.java]: Unrelated change, revert!

File: maven-aether-provider/src/main/java/org/apache/maven/repository/internal/DefaultVersionRangeResolver.java [hunk 5]
```diff
@@ -88,19 +91,27 @@
...
+            @Nullable VersionRangeResultFilter versionRangeResultFilter )
```
Comment (michael-o) [inline maven-aether-provider/src/main/java/org/apache/maven/repository/internal/DefaultVersionRangeResolver.java]: Indent as above.

File: maven-aether-provider/src/main/java/org/apache/maven/repository/internal/DefaultVersionRangeResolver.java [hunk 6]
```diff
@@ -136,15 +152,22 @@ public DefaultVersionRangeResolver setSyncContextFactory( SyncContextFactory syn
     }
 
     public DefaultVersionRangeResolver setRepositoryEventDispatcher(
-        RepositoryEventDispatcher repositoryEventDispatcher )
+            RepositoryEventDispatcher repositoryEventDispatcher )
```
Comment (michael-o) [inline maven-aether-provider/src/main/java/org/apache/maven/repository/internal/DefaultVersionRangeResolver.java]: Unrelated change, revert!

File: maven-aether-provider/src/main/java/org/apache/maven/repository/internal/DefaultVersionRangeResolver.java [hunk 7]
```diff
@@ -136,15 +152,22 @@ public DefaultVersionRangeResolver setSyncContextFactory( SyncContextFactory syn
...
     {
         this.repositoryEventDispatcher = Validate.notNull( repositoryEventDispatcher,
-            "repositoryEventDispatcher cannot be null" );
+                "repositoryEventDispatcher cannot be null" );
```
Comment (michael-o) [inline maven-aether-provider/src/main/java/org/apache/maven/repository/internal/DefaultVersionRangeResolver.java]: Unrelated change, revert!

File: maven-aether-provider/src/main/java/org/apache/maven/repository/internal/DefaultVersionRangeResolver.java [hunk 8]
```diff
@@ -136,15 +152,22 @@ public DefaultVersionRangeResolver setSyncContextFactory( SyncContextFactory syn
...
+    public DefaultVersionRangeResolver setVersionRangeResultFilter( VersionRangeResultFilter versionRangeResultFilter )
+    {
+        this.versionRangeResultFilter = Validate.notNull( versionRangeResultFilter,
+                "versionRangeResultFilter cannot be null" );
         return this;
     }
 
     public VersionRangeResult resolveVersionRange( RepositorySystemSession session, VersionRangeRequest request )
-        throws VersionRangeResolutionException
+            throws VersionRangeResolutionException
```
Comment (michael-o) [inline maven-aether-provider/src/main/java/org/apache/maven/repository/internal/DefaultVersionRangeResolver.java]: Unrelated change, revert!

File: maven-aether-provider/src/main/java/org/apache/maven/repository/internal/DefaultVersionRangeResolver.java [hunk 9]
```diff
@@ -193,19 +216,18 @@ public VersionRangeResult resolveVersionRange( RepositorySystemSession session,
             result.setVersions( versions );
         }
 
-        return result;
+        return versionRangeResultFilter.filterVersionRangeResult( result );
     }
 
     private Map<String, ArtifactRepository> getVersions( RepositorySystemSession session, VersionRangeResult result,
-                                                         VersionRangeRequest request )
+            VersionRangeRequest request )
```
Comment (michael-o) [inline maven-aether-provider/src/main/java/org/apache/maven/repository/internal/DefaultVersionRangeResolver.java]: Unrelated change, revert!

File: maven-aether-provider/src/main/java/org/apache/maven/repository/internal/DefaultVersionRangeResolver.java [hunk 10]
```diff
@@ -193,19 +216,18 @@ public VersionRangeResult resolveVersionRange( RepositorySystemSession session,
...
     {
         RequestTrace trace = RequestTrace.newChild( request.getTrace(), request );
 
         Map<String, ArtifactRepository> versionIndex = new HashMap<>();
 
-        Metadata metadata =
-            new DefaultMetadata( request.getArtifact().getGroupId(), request.getArtifact().getArtifactId(),
-                                 MAVEN_METADATA_XML, Metadata.Nature.RELEASE_OR_SNAPSHOT );
+        Metadata metadata = new DefaultMetadata( request.getArtifact().getGroupId(),
+                request.getArtifact().getArtifactId(), MAVEN_METADATA_XML, Metadata.Nature.RELEASE_OR_SNAPSHOT );
```
Comment (michael-o) [inline maven-aether-provider/src/main/java/org/apache/maven/repository/internal/DefaultVersionRangeResolver.java]: Unrelated change, revert!

File: maven-aether-provider/src/main/java/org/apache/maven/repository/internal/DefaultVersionRangeResolver.java [hunk 11]
```diff
@@ -255,7 +277,7 @@ public VersionRangeResult resolveVersionRange( RepositorySystemSession session,
     }
 
     private Versioning readVersions( RepositorySystemSession session, RequestTrace trace, Metadata metadata,
-                                     ArtifactRepository repository, VersionRangeResult result )
+            ArtifactRepository repository, VersionRangeResult result )
```
Comment (michael-o) [inline maven-aether-provider/src/main/java/org/apache/maven/repository/internal/DefaultVersionRangeResolver.java]: Unrelated change, revert!

File: maven-aether-provider/src/main/java/org/apache/maven/repository/internal/DefaultVersionRangeResolver.java [hunk 12]
```diff
@@ -290,7 +312,7 @@ private Versioning readVersions( RepositorySystemSession session, RequestTrace t
     }
 
     private void invalidMetadata( RepositorySystemSession session, RequestTrace trace, Metadata metadata,
-                                  ArtifactRepository repository, Exception exception )
+            ArtifactRepository repository, Exception exception )
```
Comment (michael-o) [inline maven-aether-provider/src/main/java/org/apache/maven/repository/internal/DefaultVersionRangeResolver.java]: Unrelated change, revert!

**Comment by barthel** [comment]
@michael-o Indent changes reverted.

**Comment by michael-o** [comment]
@barthel Very good. Did you have a chance to work on the other issues I have written?

**Comment by barthel** [comment]
@michael-o I'm working on integration tests right now.

**Comment by barthel** [comment]
@michael-o Done.

**Comment by michael-o** [comment]
I am already working on your PRs. I think, committing this under MNG-3092 is wrong and will explain why. I we cannot close the issue, people won't see this in the changelog. They won't simply know that a simple extension can solve their problem. Can you create a separate issue for your solution and make MNG-3092 depend on your issue. People would then know that they can solve it by adding an extension. I will change the PR content according to the new issue id.

**Comment by michael-o** [comment]
@barthel ITs pass. I am now willing to merge both PRs. Just waiting for your comment on the previous message.

**Comment by barthel** [comment]
@michael-o Please take a look at https://issues.apache.org/jira/browse/MNG-6049

**Comment by michael-o** [comment]
@barthel Perfect, already assigned. Just waiting for an updated PR and all will be merged.

**Comment by barthel** [comment]
Relabelling commit from MNG-3092 to MNG-6049.

**Comment by talios** [comment]
@barthel is there an example anywhere of using this?  I have a project (custom lifecycle extension) myself where I think utilizing this will actually make what I'm doing much easier, nicer, and probably more "maven safe".

**Comment by barthel** [comment]
@talios Not yet public. I'll create a public example extension if the 3.4.0 release is available.
See integration test: https://github.com/apache/maven-integration-testing/pull/14
Not easy to read, but maybe enough for you?

**Comment by talios** [comment]
@barthel Maybe, I'm running 3.4.0-SNAPSHOTs anyway, and might pull in your PR to play with and see if it helps with what I'm doing.  My currently problem is that IntelliJ IDEA doesn't seem to execute any lifecycle extensions, so things that affect dependency resolution don't get reflected in IntelliJ ( my process is currently injecting user properties, but I suspect a pluggable dependency selector may also not get picked up ).

**Comment by jira-importer** [comment]
Resolve #4815

**Comment by jira-importer** [comment]
Resolve #4815

**Comment by jira-importer** [comment]
Resolve #7699

**Comment by jira-importer** [comment]
Resolve #8061

**Comment by jira-importer** [comment]
Resolve #8616

**Comment by jira-importer** [comment]
Resolve #8616
